### PR TITLE
[mlir][sparse] introduce `sparse_tensor.coiterate` operation.

### PR DESCRIPTION
**Pull request by PeimingLiu**
This PR introduces `sparse_tensor.coiterate` operation, which represents a loop that traverses multiple sparse iteration space.

### CODE REVIEW

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td [hunk 1]
```diff
@@ -1644,6 +1644,123 @@ def IterateOp : SparseTensor_Op<"iterate",
   let hasCustomAssemblyFormat = 1;
 }
 
+def SparseTensor_CoIterateOp : SparseTensor_Op<"coiterate",
+    [AttrSizedOperandSegments,
+     SingleBlockImplicitTerminator<"sparse_tensor::YieldOp">,
+     RecursiveMemoryEffects]> {
+  let summary = "CoIterates over a set of sparse iteration spaces";
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td:1651]: Co-iterates over ...

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td [hunk 2]
```diff
@@ -1644,6 +1644,123 @@ def IterateOp : SparseTensor_Op<"iterate",
...
+  let description = [{
+      The `sparse_tensor.coiterate` operation represents a loop (nest) over
+      the a set of iteration spaces.
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td:1654]: the a -> a

Also, don't leave so much space, just put 

The operation


on this line also

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td [hunk 3]
```diff
@@ -1644,6 +1644,123 @@ def IterateOp : SparseTensor_Op<"iterate",
...
+      case %it1, _ {
+        // %coord is specifed in space %sp1 but *NOT* specified in space %sp2.
+      }
+      case %it1, %it2 {
+        // %coord is specifed in *BOTH* spaces %sp1 and %sp2.
+      }
+      ```
+
+      `sparse_tensor.coiterate` can also operate on loop-carried variables.
+      It returns the final values after loop termination.
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td:1673]: The final value   (singular)

or, if you want to make it clear there can be more

It returns the final value for each loop-carried variable after loop termination.

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td [hunk 4]
```diff
@@ -1644,6 +1644,123 @@ def IterateOp : SparseTensor_Op<"iterate",
...
+          .slice(getCrdUsedLvls().count(), getNumRegionIterArgs(regionIdx));
+    }
+    Block::BlockArgListType getRegionIterators(unsigned regionIdx) {
+      return getRegion(regionIdx).getArguments()
+          .take_back(getRegionDefinedSpace(regionIdx).count());
+    }
+  }];
+
+  // TODO:
+  // let hasVerifier = 1;
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td:1758]: don't forget to add invalid.mlir tests as well when you add the verifiers

**Review by ingomueller-net — COMMENTED** [review]
Nice progress!

It's still WIP, right? At least there's a TODO that looks like it'll be addressed as part of this PR, right?

Here and in previous PRs, a high-level summary in the PR description would have helped me to see the big picture faster.

**Comment by aartbik** [comment]
Have a look at Ingo's feedback. Keeping the TODO is okay with me, since you plan to immediately follow up with a next PR on filling out those details.

**Review by aartbik — APPROVED** [review]

**Pull request merged**